src/net/neighbors.rs

```diff
@@ -0,0 +1,2557 @@
+                        self.result.add_broken(self.cur_neighbor.addr.clone());
+                        Err(net_error::InvalidMessage)
+                    }
+                }
+            },
+            Err(req_res) => {
+                match req_res {
+                    Ok(same_req) => {
+                        // try again
+                        self.getneighbors_request = Some(same_req);
+                        Ok(None)
+                    },
+                    Err(e) => {
+                        // disconnected 
+                        self.result.add_broken(self.cur_neighbor.addr.clone());
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Begin getting the neighors of cur_neighbor's neighbors.
+    /// NetworkReplyHandles should be reply handles for Handshake requests.
+    pub fn neighbor_handshakes_begin(&mut self, local_peer: &LocalPeer, mut handshake_handles: HashMap<NeighborAddress, NetworkReplyHandle>) -> () {
+        assert!(self.state == NeighborWalkState::GetHandshakesBegin);
+
+        // advance state!
+        self.unresolved_handshake_neighbors.clear();
+        for (naddr, nh) in handshake_handles.drain() {
+            self.unresolved_handshake_neighbors.insert(naddr, nh);
+        }
+
+        self.set_state(local_peer, NeighborWalkState::GetHandshakesFinish);
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(conn: &DBConn, nk: &NeighborKey) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(|_e| net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+        
+        for slot in slots {
+            let peer_opt = PeerDB::get_peer_at(conn, nk.network_id, slot)
+                .map_err(|_e| net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
+                Some(_) => {
+                    return Ok(Some(slot));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+
+    /// Try to finish getting handshakes from cur_neighbors' neighbors.
+    /// Once all handles resolve, return the list of neighbors that we can contact.
+    /// As a side-effect of handshaking with all these peers, our PeerDB instance will be expanded
+    /// with the addresses, public keys, public key expiries of these neighbors -- i.e. this method grows
+    /// our frontier.
+    pub fn neighbor_handshakes_try_finish<'a>(&mut self, tx: &mut Transaction<'a>, local_peer: &LocalPeer, block_height: u64) -> Result<Option<Vec<NeighborKey>>, net_error> {
```

Comment:
What happens if a potential peer keeps the TCP connection open, but never sends data (try_recv will return an Empty error)? It seems like the peer will always get readded to the unresolved_handshake_neighbors set (and preventing the frontier from ever expanding).

Reply:
The `try_recv()` method will eventually fail due to a timeout.  The call to `send_message(...)` does two things:  (1) it instantiates a `NetworkReplyHandle` for the caller to wait on, and (2) it sets up a corresponding `ReceiverNotify` struct in the underlying `Connection`'s `ConnectionOutbox` that will get used to awaken the `NetworkReplyHandle` when either a reply arrives or the request times out.

The timeouts are handled in the main loop in `p2p.rs`, in `dispatch_network()`.  In this method, it will call each ongoing `Conversation`'s `clear_timeouts()` method, which will inspect the `ttl` of each `ReceiverReply` and if it exceeds the current time, it will destroy it.  This causes the `SyncSender` channel in the `ReceiverReply` to be dropped, which will cause the next call to `try_recv()` in the `NetworkReplyHandle` to fail with `Err(net_error::ConnectionBroken)`.